app/src/main/java/br/com/drss/pokedex/home/repository/PokemonRepository.kt

```diff
@@ -0,0 +1,76 @@
+package br.com.drss.pokedex.home.repository
+
+import br.com.drss.pokedex.home.repository.database.PokemonSummaryDao
+import br.com.drss.pokedex.home.repository.domain.entities.PokemonSummary
+import br.com.drss.pokedex.home.repository.domain.entities.PokemonType
+import br.com.drss.pokedex.home.repository.network.PokeApi
+import br.com.drss.pokedex.home.repository.network.dto.PokemonDto
+import kotlinx.coroutines.ExperimentalCoroutinesApi
+import kotlinx.coroutines.flow.*
+
+interface PokemonRepository {
+
+    fun getPokemonSummaryList(typeFilters: List<PokemonType> = listOf()): Flow<List<PokemonSummary>>
+    suspend fun getPokemonSummary(name: String): PokemonSummary
+
+}
+
+class PokemonRepositoryImpl(
+    private val pokemonSummaryDao: PokemonSummaryDao,
+    private val pokemonApi: PokeApi
+) : PokemonRepository {
+
+    /**
+     *
+     */
+    @ExperimentalCoroutinesApi
+    override fun getPokemonSummaryList(typeFilters: List<PokemonType>): Flow<List<PokemonSummary>> =
+        channelFlow {
+            val query =
+                if (typeFilters.isNotEmpty())
+                    pokemonSummaryDao.getTypeFilteredSummaries(typeFilters)
+                else
+                    pokemonSummaryDao.getAllSummaries()
+
+            query.collect {
+                if (it.isNotEmpty()) {
+                    offer(it)
+                } else {
+                    pokemonApi.getPokemonPagedList().results
+                        .map { getPokemonSummary(it.name) }
+                        .filter { typeFilters.isEmpty() || it.types.intersect(typeFilters).isNotEmpty() }
+                        .run { offer(this) }
+                }
+            }
+        }
+
+    /**
+     * Returns the Pokemon Summary given the [name] provided.
+     *
+     * The method will first try fetching from the local database, if nothing is found
+     * the search will occur on the API.
+     *
+     * @param [name] the name of the Pokemon.
+     *
+     * returns: PokemonSummary
+     */
+    override suspend fun getPokemonSummary(name: String): PokemonSummary {
```

Comment:
I would argue about having `pokemonSummaryDao ` and `pokemonApi ` being passes as a parameter. Since this function is a child of getPokemonSummaryList